### PR TITLE
Add various pathtracer render engine controls

### PR DESCRIPTION
**Pull request by Mango-3**
Add renderer controls for:

- Direct Lighting on/off
- Ray type bounce limits for diffuse, glossy, specular, volume
- Volume Distance Samples
- Russian Roulette Start Bounce
- Optimize for Light Outside Volumes
- Roughness Clamping
- Low Light Threshold
- Direct and IBL samples multiplier
- Embree on/off

### CODE REVIEW

**Comment by usakhelo** [comment]
This is only interface part, right?

File: src/appleseed-max-impl/appleseedrenderer/renderersettings.cpp [hunk 1]
```diff
@@ -58,9 +58,25 @@ namespace
 
             m_gi = true;
             m_caustics = false;
-            m_bounces = 3;
+            m_bounces = 8;
+            m_dbounces = 3;
```
Comment (usakhelo) [inline src/appleseed-max-impl/appleseedrenderer/renderersettings.cpp]: I would probably use more complete name, like `m_max_diffuse_bounces`. Or at least `m_diffuse_bounces".

File: src/appleseed-max-impl/appleseedrenderer/renderersettings.cpp [hunk 2]
```diff
@@ -58,9 +58,25 @@ namespace
...
+            m_dbounces_set = false;
```
Comment (usakhelo) [inline src/appleseed-max-impl/appleseedrenderer/renderersettings.cpp]: And probably `m_diffuse_bounces_enabled` here.

File: src/appleseed-max-impl/appleseedrenderer/renderersettings.cpp (hunk 1 above)
Comment (Mango-3) [reply src/appleseed-max-impl/appleseedrenderer/renderersettings.cpp]: `m_bounces` was there already, I have not introduced it. It refers to the global bounce limit. I changed the default to 8, initially it was set to 3 which is too low. 
But I agree that with more settings getting added we should make some of them more verbose.

File: src/appleseed-max-impl/appleseedrenderer/renderersettings.cpp (hunk 2 above)
Comment (Mango-3) [reply src/appleseed-max-impl/appleseedrenderer/renderersettings.cpp]: ok, no problem

**Comment by Mango-3** [comment]
> This is only interface part, right? I don't see anything saved/loaded in renderersettings.cpp. It's been decided that we skip saving in renderersettings?

> Also I don't see these new parameters referred anywhere in renderer.. I'm missing something obvious I'm sure.

The new settings are saved, both when saving the scene as Max binary file and when exporting to an appleseed scene file.

Here a scene exported to as_studio. I had set the render parameters different from defaults and you can see they are all imported into the studio scene. 
image

image

File: src/appleseed-max-impl/appleseedrenderer/datachunks.h [hunk 3]
```diff
@@ -36,35 +36,52 @@
...
-const USHORT ChunkSettingsLightingMaxRayIntensitySet    = 0x1240;
-const USHORT ChunkSettingsLightingMaxRayIntensity       = 0x1250;
-const USHORT ChunkSettingsLightingBackgroundEmitsLight  = 0x1230;
-const USHORT ChunkSettingsLightingBackgroundAlpha       = 0x1260;
-const USHORT ChunkSettingsLightingForceOffDefaultLights = 0x1270;
+const USHORT ChunkSettingsLighting                                  = 0x1200;
+const USHORT ChunkSettingsLightingGI                                = 0x1210;
+const USHORT ChunkSettingsLightingRRMinPathLength                   = 0x1211;
+const USHORT ChunkSettingsLightingCaustics                          = 0x1215;
+const USHORT ChunkSettingsLightingBounces                           = 0x1220;
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedrenderer/datachunks.h]: Should be renamed too.

File: src/appleseed-max-impl/appleseedrenderer/resource.h [hunk 4]
```diff
@@ -52,14 +79,13 @@
...
 #define IDC_SPINNER_SCALE_MULTIPLIER                409
-
 #define IDD_FORMVIEW_RENDERERPARAMS_SYSTEM          500
 #define IDC_TEXT_RENDERINGTHREADS                   501
 #define IDC_SPINNER_RENDERINGTHREADS                502
 #define IDC_CHECK_LOW_PRIORITY_MODE                 503
 #define IDC_CHECK_USE_MAX_PROCEDURAL_MAPS           504
 #define IDC_CHECK_LOG_MATERIAL_EDITOR               505
-
+#define IDC_CHECK_LOW_PRIORITY_MODE2                506
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedrenderer/resource.h]: Why do we both have `IDC_CHECK_LOW_PRIORITY_MODE` and `IDC_CHECK_LOW_PRIORITY_MODE2`?

File: src/appleseed-max-impl/appleseedrenderer/appleseedrenderer.cpp [hunk 5]
```diff
@@ -107,14 +107,30 @@ namespace
         ParamIdForceDefaultLightsOff    = 13,
         ParamIdEnableBackgroundLight    = 14,
         ParamIdBackgroundAlphaValue     = 15,
-
+        ParamIdRoughnessClamping        = 23,
+        ParamIdEnableDiffBounceLimit    = 25,
```
[the remark this answers is not in the document]
Comment (Mango-3) [reply src/appleseed-max-impl/appleseedrenderer/appleseedrenderer.cpp]: Ok, no problem

File: src/appleseed-max-impl/appleseedrenderer/appleseedrenderer.cpp [hunk 6]
```diff
@@ -604,6 +768,103 @@ ParamBlockDesc2 g_param_block_desc(
...
+        p_ui, ParamMapIdLighting, TYPE_SINGLECHEKBOX, IDC_CHECK_ENABLE_DBOUNCE,
+        p_default, FALSE,
+        p_enable_ctrls, 1, ParamIdDiffBounceLimit,
+        p_accessor, &g_pblock_accessor,
+    p_end,
+
+    ParamIdDiffBounceLimit, L"diffuse_bounce_limit", TYPE_INT, P_TRANSIENT, 0,
+        p_ui, ParamMapIdLighting, TYPE_SPINNER, EDITTYPE_INT, IDC_TEXT_DBOUNCES, IDC_SPINNER_DBOUNCES, SPIN_AUTOSCALE,
+        p_default, 3,
+        p_range, 0, 200,
```
[the remark this answers is not in the document]
Comment (Mango-3) [reply src/appleseed-max-impl/appleseedrenderer/appleseedrenderer.cpp]: I changed all Max bounce limits to 100

File: src/appleseed-max-impl/appleseedrenderer/appleseedrenderer.cpp [hunk 7]
```diff
@@ -604,6 +768,103 @@ ParamBlockDesc2 g_param_block_desc(
...
+        p_ui, ParamMapIdLighting, TYPE_SPINNER, EDITTYPE_INT, IDC_TEXT_VBOUNCES, IDC_SPINNER_VBOUNCES, SPIN_AUTOSCALE,
+        p_default, 8,
+        p_range, 0, 200,
+        p_accessor, &g_pblock_accessor,
+    p_end,
+
+    ParamIdVolumeDistanceSamples, L"volume_distance_samples", TYPE_INT, P_TRANSIENT, 0,
+        p_ui, ParamMapIdLighting, TYPE_SPINNER, EDITTYPE_INT, IDC_TEXT_VOL_DISTANCE_SAMPLES, IDC_SPINNER_VOL_DISTANCE_SAMPLES, SPIN_AUTOSCALE,
+        p_default, 2,
+        p_range, 0, 100,
```
[the remark this answers is not in the document]
Comment (Mango-3) [reply src/appleseed-max-impl/appleseedrenderer/appleseedrenderer.cpp]: ok

File: src/appleseed-max-impl/appleseedrenderer/appleseedrenderer.cpp [hunk 8]
```diff
@@ -604,6 +768,103 @@ ParamBlockDesc2 g_param_block_desc(
...
+        p_accessor, &g_pblock_accessor,
+    p_end,
+
+    ParamIdOptLightOutsideVolumes, L"optimize_for_lights_outside_volumes", TYPE_BOOL, P_TRANSIENT, 0,
+        p_ui, ParamMapIdLighting, TYPE_SINGLECHEKBOX, IDC_CHECK_LIGHT_OUTSIDE_VOLUMES,
+        p_default, FALSE,
+        p_accessor, &g_pblock_accessor,
+    p_end,
+
+    ParamIdDirectLightSamples, L"direct_light_samples", TYPE_INT, P_TRANSIENT, 0,
```
[the remark this answers is not in the document]
Comment (Mango-3) [reply src/appleseed-max-impl/appleseedrenderer/appleseedrenderer.cpp]: That's interesting, didn't know this. I set the new limits.

File: src/appleseed-max-impl/appleseedrenderer/appleseedrenderer.cpp [hunk 9]
```diff
@@ -604,6 +768,103 @@ ParamBlockDesc2 g_param_block_desc(
...
+        p_ui, ParamMapIdLighting, TYPE_SPINNER, EDITTYPE_INT, IDC_TEXT_DL_SAMPLES, IDC_SPINNER_DL_SAMPLES, SPIN_AUTOSCALE,
+        p_default, 1,
+        p_range, 1, 20,
+        p_accessor, &g_pblock_accessor,
+    p_end,
+
+    ParamIdLowLightThreshold, L"low_light_threshold", TYPE_FLOAT, P_TRANSIENT, 0,
+        p_ui, ParamMapIdLighting, TYPE_SPINNER, EDITTYPE_FLOAT, IDC_TEXT_DL_LOW_LIGHT_THRESHOLD, IDC_SPINNER_DL_LOW_LIGHT_THRESHOLD, SPIN_AUTOSCALE,
+        p_default, 0.00f,
+        p_range, 0.00f, 100.00f,
```
[the remark this answers is not in the document]
Comment (Mango-3) [reply src/appleseed-max-impl/appleseedrenderer/appleseedrenderer.cpp]: ok

File: src/appleseed-max-impl/appleseedrenderer/appleseedrenderer.rc [hunk 10]
```diff
@@ -87,27 +87,65 @@ BEGIN
...
+    CONTROL         "Background Alpha",IDC_TEXT_BACKGROUND_ALPHA,"CustEdit",WS_TABSTOP,172,198,19,10
     CONTROL         "Background Alpha",IDC_SPINNER_BACKGROUND_ALPHA,
-                    "SpinnerControl",WS_TABSTOP,93,79,6,10
+                    "SpinnerControl",WS_TABSTOP,193,198,6,10
+    CONTROL         "Roughness Clamping",IDC_CHECK_ROUGHNESS,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,105,166,84,10
+    CONTROL         "Max Diffuse Bounces:",IDC_CHECK_ENABLE_DBOUNCE,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,0,50,85,10
+    CONTROL         "Max Glossy Bounces:",IDC_CHECK_ENABLE_GBOUNCE,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,0,66,84,10
+    CONTROL         "Max Specular Bounces:",IDC_CHECK_ENABLE_SBOUNCE,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,0,82,91,10
+    CONTROL         "Max Volume Bounces:",IDC_CHECK_ENABLE_VBOUNCE,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,0,98,87,10
+    CONTROL         "Max. Bounces",IDC_TEXT_DBOUNCES,"CustEdit",WS_TABSTOP,105,50,16,10
```
[the remark this answers is not in the document]
Comment (Mango-3) [reply src/appleseed-max-impl/appleseedrenderer/appleseedrenderer.rc]: I changed them.

File: src/appleseed-max-impl/appleseedrenderer/datachunks.h (hunk 3 above)
Comment (Mango-3) [reply src/appleseed-max-impl/appleseedrenderer/datachunks.h]: ok

File: src/appleseed-max-impl/appleseedrenderer/resource.h (hunk 4 above)
Comment (Mango-3) [reply src/appleseed-max-impl/appleseedrenderer/resource.h]: This comes when one copies and pastes an UI-element in the resource view, e.g. a spinner element for a new parameter. VS Studio then automatically adds a corresponding entry for it in resource.h. When the UI-element name is then updated to fit the new parameter the old entry in resource.h still remains and can be safely deleted.  
I deleted the entry.

File: src/appleseed-max-impl/appleseedrenderer/renderersettings.cpp [hunk 11]
```diff
@@ -56,7 +56,7 @@ namespace
             m_pixel_filter = 0;
             m_pixel_filter_size = 1.5f;
 
-            m_gi = true;
+            m_enable_gi = true;
             m_caustics = false;
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedrenderer/renderersettings.cpp]: This should be called `m_enable_caustics` then.

File: src/appleseed-max-impl/appleseedrenderer/appleseedrenderer.cpp [hunk 12]
```diff
@@ -89,48 +89,50 @@ namespace
...
+
+        ParamIdCPUCores                     = 16,
+        ParamIdOpenLogMode                  = 17,
+        ParamIdLogMaterialRendering         = 18,
+        ParamIdUseMaxProcedurals            = 19,
+        ParamIdEnableLowPriority            = 20,
+        ParamIdEnableRenderStamp            = 21,
+        ParamIdRenderStampFormat            = 22,
+        ParamIdEnableEmbree                 = 24,
+
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedrenderer/appleseedrenderer.cpp]: Remove extra blank line.

**Pull request merged**